### PR TITLE
feat(checks)!: Stop running checks on every sub report of a cv report

### LINKED ISSUE #3054: `CrossValidationReport`: Stop running every check on every sub `EstimatorReport`
This makes it quite slow to run checks on cv reports. Some checks should run at the cv report level directly, e.g. SKD002 (overfitting) could build a dummy cv report and compare the mean metric values.

Two options to solve this:
- have a separate path in each relevant check that depends on the report type.
- create a cv report version of each relevant check.

I prefer option 2 as it requires less changes to the general piping of the checks (for instance, checks currently can only accept one report type). Although for option two it will require finding a way to skip estimator report checks that have a cv counterpart.

### PR DESCRIPTION
**Pull request by GaetandeCast**
Closes #3054

### CODE REVIEW

File: skore/tests/unit/reports/cross_validation/test_checks.py [hunk 1]
```diff
@@ -28,28 +39,70 @@ def check_function(self, report):
...
-    """Not-applicable split checks are surfaced on the CV report summary."""
+def test_cv_passed_checks_appear_in_summarize(regression_data):
+    """Passed CV-scoped checks without findings appear on the CV report summary."""
     X, y = regression_data
-    report = evaluate(DecisionTreeRegressor(random_state=0), X, y, splitter=3)
-    na_codes = set(report.checks.summarize().frame(section="not_applicable")["code"])
-    assert "SKD014" in na_codes
-    assert "SKD015" in na_codes
+    report = evaluate(LinearRegression(), X, y, splitter=3)
+    passed_codes = set(report.checks.summarize().frame(section="passed")["code"])
```
Comment (auguste-probabl) [inline skore/tests/unit/reports/cross_validation/test_checks.py:53]:
```suggestion
    passed_codes = report.checks.summarize().frame(section="passed")["code"]
```

File: skore/tests/unit/reports/cross_validation/test_checks.py [hunk 2]
```diff
@@ -28,28 +39,70 @@ def check_function(self, report):
...
+    assert "SKD003" in passed_codes
 
 
-def test_cv_passed_checks_appear_in_summarize(regression_data):
-    """Passed split checks without findings appear on the CV report summary."""
+def test_skd001_detects_overfitting(regression_data):
+    """Check that the overfitting issue is detected."""
```
Comment (auguste-probabl) [inline skore/tests/unit/reports/cross_validation/test_checks.py:58]:
```suggestion
    """The overfitting issue is detected properly."""
```
For consistency with the test above

File: skore/tests/unit/reports/cross_validation/test_checks.py [hunk 3]
```diff
@@ -28,28 +39,70 @@ def check_function(self, report):
...
+        ("pandas", "series"),
+        ("polars", "polars_series"),
+    ],
+)
+def test_skd002_detects_underfitting(regression_data, x_container, y_container):
+    """Check that the underfitting issue is detected."""
+    X, y = regression_data
+    feature_columns = [str(i) for i in range(X.shape[1])]
+    X = convert_container(
+        X, x_container, column_names=feature_columns, minversion="0.20.23"
```
Comment (auguste-probabl) [inline skore/tests/unit/reports/cross_validation/test_checks.py]:
```suggestion
        X, x_container, column_names=feature_columns
```
i wonder if this still passes?

File: skore/tests/unit/reports/cross_validation/test_checks.py [hunk 4]
```diff
@@ -92,20 +136,26 @@ def test_skd001_aggregates_overfitting_across_splits(regression_data):
...
+        weights=weights,
+        random_state=0,
+    )
     feature_columns = [str(i) for i in range(X.shape[1])]
     X = convert_container(
         X, x_container, column_names=feature_columns, minversion="0.20.23"
     )
     y = convert_container(y, y_container, minversion="0.20.23")
-    report = evaluate(DummyRegressor(), X, y, splitter=3)
+    report = evaluate(LogisticRegression(), X, y, splitter=0.2)
```
Comment (auguste-probabl) [inline skore/tests/unit/reports/cross_validation/test_checks.py]: this file is for CVReports but this is an EstimatorReport; i guess because this check only looks at the data?

File: skore/tests/unit/reports/cross_validation/test_checks.py (hunk 1 above)
Comment (auguste-probabl) [reply skore/tests/unit/reports/cross_validation/test_checks.py:53]: same everywhere

File: skore/tests/unit/reports/cross_validation/test_checks.py [hunk 5]
```diff
@@ -132,18 +183,233 @@ def test_skd004_detects_high_class_imbalance(x_container, y_container):
...
+    issues = report.checks.summarize().frame(section="issue").set_index("code")
+    assert "SKD014" in issues.index
+    assert "minimum" in issues.loc["SKD014", "explanation"]
+
+
+def test_skd014_not_applicable_for_plain_estimator_on_cv_report(regression_data):
+    """SKD014 raises CheckNotApplicable with a plain estimator."""
+    X, y = regression_data
+    report = evaluate(LinearRegression(), X, y, splitter=3)
+    with pytest.raises(CheckNotApplicable):
```
Comment (auguste-probabl) [inline skore/tests/unit/reports/cross_validation/test_checks.py]: can you check the error message too?

File: skore/src/skore/_sklearn/_base.py [hunk 6]
```diff
@@ -85,7 +85,7 @@ def _get_results(
         checks_to_run = [
             check
             for check in self._checks_registry
-            if check.report_type == self._report_type
+            if self._report_type in check.report_type
             and check.code not in self._check_results_cache
             and check.code not in ignored_codes
             and not (fast_mode and getattr(check, "slow", False))
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_base.py]:
```suggestion
            and not (fast_mode and check.slow)
```

File: skore/src/skore/_sklearn/_checks/model_checks.py [hunk 7]
```diff
@@ -594,65 +599,93 @@ class CheckGoldenFeature(Check):
...
-        X_test = cast("UserDataFrame", get_preprocessed_X(report, data_source="test"))
-        y_train = get_report_y(report, data_source="train")
-        y_test = cast("UserTarget", get_report_y(report, data_source="test"))
-        if X_train is None or y_train is None:
-            raise CheckNotApplicable("Train data is unavailable.")
-        if X_train.shape[1] < 2:
-            raise CheckNotApplicable("Train data has only one feature.")
-
-        preprocessor_, predictor_ = split_preprocessor_estimator(report.estimator_)
+        report = cast_report(report)
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_checks/model_checks.py]:
```suggestion
```

File: skore/src/skore/_sklearn/_checks/model_checks.py [hunk 8]
```diff
@@ -594,65 +599,93 @@ class CheckGoldenFeature(Check):
...
+        is_cv = report._report_type == "cross-validation"
+        if is_cv:
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_checks/model_checks.py]:
```suggestion
        if report._report_type == "cross-validation":
```

File: skore/src/skore/_sklearn/_checks/model_checks.py [hunk 9]
```diff
@@ -594,65 +599,93 @@ class CheckGoldenFeature(Check):
...
+            cv_report = cast("CrossValidationReport", report)
+            X = nw.from_native(get_preprocessed_X(cv_report))
+            y = get_report_y(cv_report)
+            if X.shape[1] < 2:
+                raise CheckNotApplicable("Train data has only one feature.")
+            n_features = X.shape[1]
+            metric_registry = cv_report.reports_[0]._metric_registry.copy()
+            from skore._sklearn._cross_validation.report import CrossValidationReport
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_checks/model_checks.py:616]:
```suggestion
```

File: skore/src/skore/_sklearn/_checks/model_checks.py [hunk 10]
```diff
@@ -594,65 +599,93 @@ class CheckGoldenFeature(Check):
...
+            X_test = nw.from_native(
+                get_preprocessed_X(estimator_report, data_source="test")
+            )
+            y_train = get_report_y(estimator_report, data_source="train")
+            y_test = get_report_y(estimator_report, data_source="test")
+            if X_train.shape[1] < 2:
+                raise CheckNotApplicable("Train data has only one feature.")
+            n_features = X_train.shape[1]
+            metric_registry = estimator_report._metric_registry
+            from skore._sklearn._estimator.report import EstimatorReport
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_checks/model_checks.py:627]:
```suggestion
```

File: skore/src/skore/_sklearn/_checks/model_checks.py [hunk 11]
```diff
@@ -730,38 +763,52 @@ class CheckTrainTestTimeOverlap(Check):
...
+        ):
+            if report.X_train is None:
+                raise CheckNotApplicable("Train data is unavailable.")
+            if not nw.dependencies.is_into_dataframe(
+                report.X_train
+            ) or not nw.dependencies.is_into_dataframe(report.X_test):
+                raise CheckNotApplicable(
+                    "Input data is not a narwhals compatible DataFrame. "
+                    f"Got {type(report.X_train)}."
+                )
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_checks/model_checks.py]: if report.X_test causes the error, this message will be misleading.

```suggestion
            if not nw.dependencies.is_into_dataframe(report.X_train):
                raise CheckNotApplicable(
                    "Input data is not a narwhals compatible DataFrame. "
                    f"Got {type(report.X_train)}."
                )
            if not nw.dependencies.is_into_dataframe(report.X_test):
                raise CheckNotApplicable(
                    "Input data is not a narwhals compatible DataFrame. "
                    f"Got {type(report.X_test)}."
                )
```

Comment (auguste-probabl) [reply skore/src/skore/_sklearn/_checks/model_checks.py]: or `try-except` the `from_native` calls

File: skore/src/skore/_sklearn/_checks/base.py [hunk 12]
```diff
@@ -271,7 +272,7 @@ class Check(Protocol):
 
     code: CheckCode
     title: str
-    report_type: ReportType
+    report_type: list[ReportType] | tuple[ReportType]
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_checks/base.py]:
```suggestion
    report_type: list[ReportType]
```
per the docstring

File: skore/src/skore/_sklearn/_checks/base.py [hunk 13]
```diff
@@ -252,9 +252,10 @@ class Check(Protocol):
         is shown as-is; otherwise it is treated as an HTML anchor fragment under
         the automated checks user guide.
 
-    report_type : str
-        Must be one of ``"cross-validation"``, ``"estimator"``,
-        ``"comparison-estimator"``, or ``"comparison-cross-validation"``.
+    report_type : list of str
+        Report types this check applies to. Each element must be one of
+        ``"cross-validation"``, ``"estimator"``, ``"comparison-estimator"``,
+        or ``"comparison-cross-validation"``.
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_checks/base.py]:
```suggestion
    report_type : list of {"estimator", "cross-validation", \
            "comparison-estimator", "comparison-cross-validation"}
        Report types this check applies to.
```

File: skore/src/skore/_sklearn/_checks/base.py (hunk 12 above)
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_checks/base.py]: Should be renamed to `report_types` or `supported_report_types`

File: skore/src/skore/_sklearn/_checks/accessor.py [hunk 14]
```diff
@@ -94,19 +94,29 @@ def add(
...
+        valid_report_types = {
             "cross-validation",
             "estimator",
             "comparison-estimator",
             "comparison-cross-validation",
-        ]
+        }
         for check in checks:
             if not isinstance(check, Check):
                 raise ValueError(f"{check} does not implement the Check protocol.")
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_checks/accessor.py]:
```suggestion
                raise TypeError(f"{check} is not a subclass of Check.")
```

File: skore/src/skore/_sklearn/_checks/accessor.py [hunk 15]
```diff
@@ -94,19 +94,29 @@ def add(
...
-            if check.report_type not in report_types:
+            if (
+                not isinstance(check.report_type, (list, tuple))
+                or not check.report_type
+            ):
+                raise ValueError(
+                    "Check report_type must be a non-empty list or tuple of report "
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_checks/accessor.py]:
```suggestion
                    "The check's report_type must be a non-empty list or tuple of report "
```

File: skore/src/skore/_sklearn/_checks/accessor.py [hunk 16]
```diff
@@ -94,19 +94,29 @@ def add(
...
+                    f"types. Got {type(check.report_type)}."
+                )
+            invalid_types = set(check.report_type) - valid_report_types
+            if invalid_types:
                 raise ValueError(
-                    f"Check report_type should be one of: {', '.join(report_types)}. "
-                    f"Got {check.report_type} instead."
+                    f"Supported values for report_type are: {valid_report_types}. "
+                    f"{check.code}'s report_type contains unsupported values: "
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_checks/accessor.py]:
```suggestion
                    "The check's report_type contains unsupported values: "
```

File: skore/src/skore/_sklearn/_checks/_utils.py [hunk 17]
```diff
@@ -44,22 +45,37 @@ def _metric_key(row: MetricsSummaryRow) -> MetricKey:
...
     return {
-        _metric_key(row): row
-        for row in rows
-        if include_timing or row["metric_verbose_name"] not in _TIMING_METRICS
+        key: cast(
+            MetricsSummaryRow,
+            {
+                **split_rows[0],
+                "score": float(np.mean([row["score"] for row in split_rows])),
+            },
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_checks/_utils.py]: nit
```suggestion
            split_rows[0] | {
                "score": float(np.mean([row["score"] for row in split_rows])),
            },
```

File: examples/technical_details/plot_custom_checks.py [hunk 18]
```diff
@@ -132,18 +131,17 @@ def check_function(self, report):
...
+# ==================================================================
 #
-# We can also reuse our first check to run it on the component estimator reports
-# and aggregate the results across splits.
-
-cv_report.checks.add([CustomCheck1()])
-cv_report.checks.summarize()
+# Estimator-scoped checks such as ``CustomCheck1`` run only on
+# :class:`~skore.EstimatorReport` instances. They are not executed on
+# :class:`~skore.CrossValidationReport`. To run the same logic on both report types,
```
Comment (auguste-probabl) [inline examples/technical_details/plot_custom_checks.py]:
```suggestion
# Estimator-scoped checks such as ``CustomCheck1`` are only executed on
# :class:`~skore.EstimatorReport` instances, not on
# :class:`~skore.CrossValidationReport` instances. To run the same logic on both report types,
```

File: examples/technical_details/plot_custom_checks.py [hunk 19]
```diff
@@ -83,13 +83,12 @@ def check_function(self, report):
...
 # :class:`~skore.CrossValidationReport` and :class:`~skore.ComparisonReport` can also
-# receive custom checks, either ran on the full report or on the component estimator
-# reports.
+# receive custom checks scoped to their report type.
 #
-# The `report_type` argument of :class:`~skore.Check` controls the scope of the check.
-# Let's write a check that is specific to cross-validation reports: it flags metrics
-# with high variance across splits. We set the severity to "issue" to indicate that
-# this is an issue to fix.
+# The `report_type` argument of :class:`~skore.Check` is a list that controls which
```
Comment (auguste-probabl) [inline examples/technical_details/plot_custom_checks.py]:
```suggestion
# The `report_type` attribute of :class:`~skore.Check` is a list that controls which
```

**Review by auguste-probabl — CHANGES_REQUESTED** [review]
a few comments

File: skore/tests/unit/reports/cross_validation/test_checks.py (hunk 1 above)
Comment (GaetandeCast) [reply skore/tests/unit/reports/cross_validation/test_checks.py:53]: Conversion to set is needed as `in pd.Series` looks at the indices not the values.

File: skore/tests/unit/reports/cross_validation/test_checks.py (hunk 2 above)
Comment (GaetandeCast) [reply skore/tests/unit/reports/cross_validation/test_checks.py:58]: The test docstrings for checks are really heterogenous, I'd rather add homogenizing them to the list of things to improve on those.

File: skore/src/skore/_sklearn/_checks/model_checks.py (hunk 9 above)
Comment (GaetandeCast) [reply skore/src/skore/_sklearn/_checks/model_checks.py:616]: This is needed. I can't import at the top otherwise it causes a circular import.

File: skore/src/skore/_sklearn/_checks/model_checks.py (hunk 10 above)
Comment (GaetandeCast) [reply skore/src/skore/_sklearn/_checks/model_checks.py:627]: same

File: skore/tests/unit/reports/cross_validation/test_checks.py (hunk 2 above)
Comment (auguste-probabl) [reply skore/tests/unit/reports/cross_validation/test_checks.py:58]: actually that's true for the whole lib

**Review by auguste-probabl — APPROVED** [review]
nice work!
